src/core/server/ui_settings/types.ts

```diff
@@ -0,0 +1,118 @@
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+import { SavedObjectsClientContract, SavedObjectAttribute } from '../saved_objects/types';
+/**
+ * Client that provides access to the UiSettings stored in elasticsearch.
+ *
+ * @public
+ */
+export interface IUiSettingsClient {
+  /**
+   * Returns registered uiSettings values {@link UiSettingsParams}
+   */
+  getRegistered: () => Readonly<Record<string, UiSettingsParams>>;
+  /**
+   * Retrieves uiSettings values set by the user with fallbacks to default values if not specified.
+   */
+  get: <T extends SavedObjectAttribute = any>(key: string) => Promise<T>;
+  /**
+   * Retrieves a set of all uiSettings values set by the user with fallbacks to default values if not specified.
+   */
+  getAll: <T extends SavedObjectAttribute = any>() => Promise<Record<string, T>>;
+  /**
+   * Retrieves a set of all uiSettings values set by the user.
+   */
+  getUserProvided: <T extends SavedObjectAttribute = any>() => Promise<
+    Record<string, { userValue?: T; isOverridden?: boolean }>
+  >;
+  /**
+   * Writes multiple uiSettings values and marks them as set by the user.
+   */
+  setMany: <T extends SavedObjectAttribute = any>(changes: Record<string, T>) => Promise<void>;
+  /**
+   * Writes uiSettings value and marks it as set by the user.
+   */
+  set: <T extends SavedObjectAttribute = any>(key: string, value: T) => Promise<void>;
+  /**
+   * Removes uiSettings value by key.
+   */
+  remove: (key: string) => Promise<void>;
+  /**
+   * Removes multiple uiSettings values by keys.
+   */
+  removeMany: (keys: string[]) => Promise<void>;
+  /**
+   * Shows whether the uiSettings value set by the user.
+   */
+  isOverridden: (key: string) => boolean;
+}
+
+/**
+ * UI element type to represent the settings.
+ * @public
+ * */
+export type UiSettingsType = 'json' | 'markdown' | 'number' | 'select' | 'boolean' | 'string';
+
+/**
+ * UiSettings parameters defined by the plugins.
+ * @public
+ * */
+export interface UiSettingsParams {
+  /** title in the UI */
+  name: string;
+  /** default value to fall back to if a user doesn't provide any */
+  value: SavedObjectAttribute;
+  /** description provided to a user in UI */
+  description: string;
+  /** used to group the configured setting in the UI */
+  category: string[];
+  /** a range of valid values */
+  options?: string[];
+  /** text labels for 'select' type UI element */
+  optionLabels?: Record<string, string>;
+  /** a flag indicating whether new value applying requires page reloading */
+  requiresPageReload?: boolean;
+  /** a flag indicating that value cannot be changed */
+  readonly?: boolean;
+  /** defines a type of UI element {@link UiSettingsType} */
```

Comment:
Is there a default?

Reply:
No. I think we assume `string` as a default. Probably we should express this in code https://github.com/restrry/kibana/blob/841abd1162f1eb6aaebf27d004c92e8344636d91/src/legacy/core_plugins/kibana/public/management/sections/settings/components/field/field.js#L494